docs/data_fetching.rst

```diff
@@ -0,0 +1,220 @@
+tweak values before sending them out.  For example above we have a ``launchDate`` field that takes an optional ``dateFormat``
+argument.  We can have the ProductDTO have logic that applies this date formatting to the desired format.
+
+
+.. code-block:: java
+
+    class ProductDTO {
+
+        private ID id;
+        private String name;
+        private String description;
+        private Double cost;
+        private Double tax;
+        private LocalDateTime launchDate;
+
+        // ...
+
+        public String getName() {
+            return name;
+        }
+
+        // ...
+
+        public String getLaunchDate(DataFetchingEnvironment environment) {
+            String dateFormat = environment.getArgument("dateFormat");
+            return yodaTimeFormatter(launchDate,dateFormat);
+        }
+    }
+
+Customising PropertyDataFetcher
+-------------------------------
+
+As mentioned above ``graphql.schema.PropertyDataFetcher`` is the default data fetcher for fields in graphql-java and it will use standard patterns for fetching
+object field values.
+
+It supports a ``POJO`` approach and a ``Map`` approach in a Java idiomatic way.  By default it assumes that for a graphql field ``fieldX`` it can find a POJO property
+called ``fieldX`` or a map key called ``fieldX`` if the backing object is a ``Map``.
+
+However you may have small differences between your graphql schema naming and runtime object naming.  For example imagine that ``Product.description`` is actually
+represented as ``getDesc()`` in the runtime backing Java object.
+
+If you are using SDL to specify your schema then you can use the ``@fetch`` directive to indicate this remapping.
+
+.. code-block:: graphql
+
+    directive @fetch(from : String!) on FIELD_DEFINITION
+
+    type Product {
+        id : ID
+        name : String
+        description : String @fetch(from:"desc")
+        cost : Float
+        tax : Float
+    }
+
+This will tell the ``graphql.schema.PropertyDataFetcher`` to use the property name ``desc`` when fetching data for the graphql field named ``description``.
+
+If you are hand coding your schema then you can just specify it directly by wiring in a field data fetcher.
+
+.. code-block:: java
+
+        GraphQLFieldDefinition descriptionField = GraphQLFieldDefinition.newFieldDefinition()
+                .name("description")
+                .type(Scalars.GraphQLString)
+                .dataFetcher(PropertyDataFetcher.fetching("desc"))
+                .build();
+
+
+
+The interesting parts of the DataFetchingEnvironment
+----------------------------------------------------
+
+Every data fetcher is passed a ``graphql.schema.DataFetchingEnvironment`` object which allows it to know more about what is being fetched
+and what arguments have been provided.  Here are some of the more interesting parts of ``DataFetchingEnvironment``.
+
+* ``<T> T getSource()`` - the ``source`` object is used to get information for a field.  In the simple case its the in memory
```

Comment:
source has always been a bit confusing name imho: I often try to describe it as the result of the parent field DF. Not sure this makes it clear, but it is technically correct.